src/content/reducers/url-state.js

```diff
@@ -159,6 +173,50 @@ function hidePlatformDetails(state: boolean = false, action: Action) {
   }
 }
 
+function threadOrder(state: ThreadIndex[] = [], action: Action) {
+  switch (action.type) {
+    case 'RECEIVE_PROFILE_FROM_ADDON':
+    case 'RECEIVE_PROFILE_FROM_STORE':
+    case 'RECEIVE_PROFILE_FROM_URL':
+    case 'RECEIVE_PROFILE_FROM_FILE': {
+      // When receiving a new profile, try to use the thread order specified in the URL,
+      // but ensure that the IDs are correct.
+      const threads = defaultThreadOrder(action.profile.threads);
+      const validURLThreads = state.filter(index => threads.includes(index));
+      const missingThreads = threads.filter(index => !state.includes(index));
+      return validURLThreads.concat(missingThreads);
+    }
+    case 'CHANGE_THREAD_ORDER':
+      return action.threadOrder;
+    default:
+      return state;
+  }
+}
+
+function hiddenThreads(state: ThreadIndex[] = [], action: Action) {
+  switch (action.type) {
+    case 'RECEIVE_PROFILE_FROM_ADDON':
+    case 'RECEIVE_PROFILE_FROM_STORE':
+    case 'RECEIVE_PROFILE_FROM_URL':
+    case 'RECEIVE_PROFILE_FROM_FILE': {
+      // When receiving a new profile, try to use the hidden threads specified in the URL,
+      // but ensure that the IDs are correct.
+      const threads = action.profile.threads.map((_, threadIndex) => threadIndex);
+      return state.filter(index => threads.includes(index));
+    }
+    case 'HIDE_THREAD': {
+      const { threadIndex } = action;
+      return [...state, threadIndex];
```

Comment:
nit: unnecessary use of destructuring, this is just as readable and doesn't need a local var: `return [...state, action.threadIndex]`